### PR TITLE
unsafe class loader and field synthesizer

### PR DESCRIPTION
**Pull request by nnym**
This pull request causes the class loader to be replaced with a subclass that is capable of loading illegally defined classes. In order to do that without causing errors, initialization is moved to `GrossFabricHacks#onLoad`. It implements synthetic field generation and access tools (mainly) for use in interfaces. It also contains a separate `Unsafe` class that abstracts the location of the `Unsafe` class for compatibility with Java 9 and above, and it cleans up a few things. Note that GrossFabricHacks is still not fully compatible with Java 9 and above due to some reflection-related fields or classes.

### CODE REVIEW

**Comment by nnym** [comment]
The time is not yet.

**Comment by nnym** [comment]
The time is now.

File: src/main/java/net/devtech/grossfabrichacks/field/FieldSynthesizer.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+import org.objectweb.asm.tree.InsnList;
+import org.objectweb.asm.tree.MethodNode;
+
+public class FieldSynthesizer {
+    private static final Logger LOGGER = LogManager.getLogger("GrossFabricHacks/FieldSynthesizer");
+
+    public static void init() {
+        LOGGER.info("Initializing field synthesis capability.");
+
+        TransformerApi.registerPreMixinAsmClassTransformer(FieldSynthesizer::transform);
```
Comment (Ueaj-Kerman) [inline src/main/java/net/devtech/grossfabrichacks/field/FieldSynthesizer.java:24]: Is there a better way of doing this that desn't need mass-asm?

File: src/main/java/net/devtech/grossfabrichacks/field/FieldSynthesizer.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        LOGGER.info("Initialization complete. Multiple inheritance of state is now at your key presses.");
+    }
+
+    private static void transform(final String name, final ClassNode klass) {
+        if (klass.visibleAnnotations != null) {
+            for (final AnnotationNode annotation : klass.visibleAnnotations) {
+                if ("Lnet/devtech/grossfabrichacks/field/Fields;".equals(annotation.desc)) {
+                    //noinspection unchecked
+                    for (final AnnotationNode field : (List<AnnotationNode>) ASMUtil.getAnnotationValue(annotation, "value")) {
+                        addField(klass, field);
```
Comment (Ueaj-Kerman) [inline src/main/java/net/devtech/grossfabrichacks/field/FieldSynthesizer.java:35]: why can't this be done lazily?

File: src/main/java/net/devtech/grossfabrichacks/field/FieldSynthesizer.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+            }
+        }
+
+        for (final MethodNode method : klass.methods) {
+            if (method.visibleAnnotations != null) {
+                for (final AnnotationNode annotation : method.visibleAnnotations) {
+                    if (Type.getDescriptor(Getter.class).equals(annotation.desc)) {
+                        FieldSynthesizer.get(name, method, annotation);
+                    } else if (Type.getDescriptor(Setter.class).equals(annotation.desc)) {
+                        FieldSynthesizer.set(name, method, annotation);
```
Comment (Ueaj-Kerman) [inline src/main/java/net/devtech/grossfabrichacks/field/FieldSynthesizer.java:47]: like here, why can't u just add fields to the class, u have the information required, so why not use it

File: src/main/java/net/devtech/grossfabrichacks/field/Fields.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.Type;
+
+/**
+ * This annotation should contain information about synthetic fields.<br>
+ * It is intended to be used with {@link Getter} or {@link Setter}.
+ */
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface Fields {
```
Comment (Ueaj-Kerman) [inline src/main/java/net/devtech/grossfabrichacks/field/Fields.java:14]: this shouldn't be needed imo, also use repeatable annotations

File: src/main/java/net/devtech/grossfabrichacks/transformer/TransformerApi.java [hunk 5]
```diff
@@ -31,6 +31,7 @@ public static void registerPreMixinRawClassTransformer(RawClassTransformer trans
 	 */
 	public static void registerPreMixinAsmClassTransformer(AsmClassTransformer transformer) {
 		TransformerBootstrap.shouldWrite = true;
+
```
Comment (Ueaj-Kerman) [inline src/main/java/net/devtech/grossfabrichacks/transformer/TransformerApi.java:34]: crob

File: src/main/java/net/devtech/grossfabrichacks/unsafe/Unsafe.java [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+            return Class.forName("jdk.internal.misc.Unsafe");
+        } catch (final ClassNotFoundException bad) {
+            try {
+                return Class.forName("sun.misc.Unsafe");
+            } catch (final ClassNotFoundException exception) {
+                throw new RuntimeException(exception);
+            }
+        }
+    }
+}
```
Comment (Ueaj-Kerman) [inline src/main/java/net/devtech/grossfabrichacks/unsafe/Unsafe.java]: add this to UnsafeUtil instead of it's own class

**Review by Ueaj-Kerman — CHANGES_REQUESTED** [review]

File: src/main/java/net/devtech/grossfabrichacks/field/FieldSynthesizer.java (hunk 3 above)
Comment (nnym) [reply src/main/java/net/devtech/grossfabrichacks/field/FieldSynthesizer.java:47]: I could do that; it's what I did before—but it turned out badly—so I came up with the present `@Fields.Entry` declaration. It also more closely resembles typical Java structure than generating fields from accessors (not to suggest that fields declared via annotations resemble typical Java code in any way).

**Review by Ueaj-Kerman — APPROVED** [review]
if it works, it works

**Pull request merged**